packages/buidler-etherscan/src/solc/bytecode.ts

```diff
@@ -0,0 +1,264 @@
+      console.log(deployedMetadataLength);
+      return { match: false };
+    }
+
+    // The metadata length is stored at the end.
+    bytecodeSize -= (deployedMetadataLength + metadataLengthSize) * 2;
+  }
+
+  // Normalize deployed bytecode according to this contract.
+  const {
+    immutableValues,
+    libraryLinks,
+    normalizedBytecode,
+  } = await normalizeBytecode(deployedBytecode, runtimeBytecodeSymbols);
+
+  // Library hash placeholders are embedded into the bytes where the library addresses are linked.
+  // We need to zero them out to compare them.
+  const { normalizedBytecode: referenceBytecode } = await normalizeBytecode(
+    runtimeBytecodeSymbols.object,
+    runtimeBytecodeSymbols
+  );
+
+  if (
+    normalizedBytecode.slice(0, bytecodeSize - 1) ===
+    referenceBytecode.slice(0, bytecodeSize - 1)
+  ) {
+    // The bytecode matches
+    return {
+      contractInformation: {
+        immutableValues,
+        libraryLinks,
+        normalizedBytecode,
+      },
+      match: true,
+    };
+  }
+
+  return { match: false };
+}
+
+interface ResolvedLinks {
+  [libraryFileGlobalName: string]: {
+    [libraryName: string]: string;
+  };
+}
+
+interface ImmutableValues {
+  [key: string]: string;
+}
+
+interface BytecodeSlice {
+  start: number;
+  length: number;
+}
+
+type LinkReferences = CompilerOutputBytecode["linkReferences"][string][string];
+type NestedSliceReferences = BytecodeSlice[][];
+
+export async function normalizeBytecode(
+  bytecode: string,
+  symbols: CompilerOutputBytecode
+) {
+  const nestedSliceReferences: NestedSliceReferences = [];
+  const libraryLinks: ResolvedLinks = {};
+  for (const [filename, libraries] of Object.entries(symbols.linkReferences)) {
+    for (const [libraryName, linkReferences] of Object.entries(libraries)) {
+      // Is this even a possibility?
+      if (linkReferences.length === 0) {
+        continue;
+      }
+
+      const { start, length } = linkReferences[0];
+      if (libraryLinks[filename] === undefined) {
+        libraryLinks[filename] = {};
+      }
+      // We have the bytecode encoded as a hex string
+      libraryLinks[filename][libraryName] = `0x${bytecode.slice(
+        start * 2,
+        (start + length) * 2
+      )}`;
+      nestedSliceReferences.push(linkReferences);
+    }
+  }
+
+  const immutableValues: ImmutableValues = {};
+  if (
+    symbols.immutableReferences !== undefined &&
+    symbols.immutableReferences !== null
+  ) {
+    for (const [key, immutableReferences] of Object.entries(
+      symbols.immutableReferences
+    )) {
+      // Is this even a possibility?
+      if (immutableReferences.length === 0) {
+        continue;
+      }
+
+      const { start, length } = immutableReferences[0];
+      immutableValues[key] = bytecode.slice(start * 2, (start + length) * 2);
+      nestedSliceReferences.push(immutableReferences);
+    }
+  }
+
+  const sliceReferences: BytecodeSlice[] = ([] as BytecodeSlice[]).concat(
```

Comment:
Two things here:

- The type annotation is not necessary
- Let's move this to a `flatten` helper for readability (or install `lodash.flatten`, I'm fine with either option)